Issue #10749 - WebSocketClient should expose upgrade request/response

lachlan-roberts opened the pull request:
## Issue #10749

Invoke `UpgradeListener.onHandshakeResponse` even on non-successful WebSocket upgrade.


Note: it is not possible to expose the core Request through `ClientUpgradeRequest` because it doesn't ever see the actual `Request`, which is created inside the call to `WebSocketClient.connect()`.

lachlan-roberts committed:
Issue #10749 - allow UpgradeListener to see Response in case of non-successful upgrade

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

sbordet commented on jetty-core/jetty-websocket/jetty-websocket-core-client/src/main/java/org/eclipse/jetty/websocket/core/client/CoreClientUpgradeRequest.java:
```diff
@@ -247,9 +247,19 @@ public void onComplete(Result result)
+        int status = response.getStatus();
+        String responseLine = status + " " + response.getReason();
+
+        if ((request.getVersion() == HttpVersion.HTTP_2 && status != HttpStatus.OK_200) ||
+            (request.getVersion() == HttpVersion.HTTP_1_1 && status != HttpStatus.SWITCHING_PROTOCOLS_101))
```
Can you replace this with a boolean field that gets set when the upgrade happens, and here just do `if (!upgraded)`?

In this way will work for HTTP/3 as well, etc.

lachlan-roberts committed:
#10761 - use boolean to track whether request was upgraded

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

sbordet committed:
Improved exception handling in HttpUpgraderOverHTTP.
Avoid using log warnings in CoreClientUpgradeRequest.
Delayed setting of this.upgraded after the last throw statement.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet reviewed (approved):
@lachlan-roberts I have pushed a commit with small changes.
If you agree on those changes, we can merge.
